README.md
koa-gzip
=======

[![Build Status](https://secure.travis-ci.org/node-modules/koa-gzip.png)](http://travis-ci.org/node-modules/koa-gzip)

[![Dependency Status](https://gemnasium.com/node-modules/koa-gzip.png)](https://gemnasium.com/node-modules/koa-gzip)

[![NPM](https://nodei.co/npm/koa-gzip.png?downloads=true&stars=true)](https://nodei.co/npm/koa-gzip/)

![logo](https://raw.github.com/node-modules/koa-gzip/master/logo.png)

`gzip` support for [koa](https://github.com/koajs/koa) responses.

## Install

```bash
$ npm install koa-gzip
```

## Usage

```js
var koa = require('koa');
var gzip = require('koa-gzip');

var app = koa();
koa.use(gzip());
```

### gzip always be the last middleware

If you using other response middlewares, like [etag](https://github.com/koajs/etag), just use `gzip` be the first.

```js
app.use(gzip());
app.use(fresh());
app.use(etag());

client => request => gzip() => fresh() => etag() => logic codes
                                                        ||
            client <= gzip() <= fresh() <= etag() <= response
```

## License

(The MIT License)

Copyright (c) 2014 fengmk2 &lt;fengmk2@gmail.com&gt; and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.1.0 / 2014-04-28
==================

 * fix links
 * Merge pull request #4 from coderhaoxin/remove-content-length
 * remove possible incorrect content-length

0.0.4 / 2014-03-23 
==================

  * if already set encoding, do not gzip (@dead-horse)

0.0.3 / 2014-03-05 
==================

  * use delegating yields http://jongleberry.com/delegating-yield.html
  * use istanbul harmony branch
  * add coverage check
  * use istanbul replace blanket. gotwarlost/istanbul#44

0.0.2 / 2014-03-03 
==================

  * add content min length == 150 bytes
  * only test on 0.11

0.0.1 / 2014-03-03 
==================

  * first commit


test/index.test.js
/**!
 * koa-gzip - test/index.test.js
 *
 * Copyright(c) 2014
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

"use strict";

/**
 * Module dependencies.
 */

var fs = require('fs');
var should = require('should');
var request = require('supertest');
var koa = require('koa');
var mm = require('mm');
var gzip = require('../');

describe('index.test.js', function () {
  var options = {};
  var BODY = 'foo bar string, foo bar string, foo bar string, foo bar string, \
    foo bar string, foo bar string, foo bar string, foo bar string, foo bar string, foo bar string, \
    foo bar string, foo bar string, foo bar string, foo bar string, foo bar string, foo bar string';
  var app = koa();
  // app.outputErrors = true;
  app.use(gzip(options));
  app.use(gzip(options));
  app.use(function *(next) {
    if (this.url === '/404') {
      return yield next;
    }
    if (this.url === '/small') {
      return this.body = 'foo bar string';
    }
    if (this.url === '/string') {
      return this.body = BODY;
    }
    if (this.url === '/buffer') {
      return this.body = new Buffer(BODY);
    }
    if (this.url === '/object') {
      return this.body = {foo: BODY};
    }
    if (this.url === '/number') {
      return this.body = 1984;
    }
    if (this.url === '/stream') {
      var stat = fs.statSync(__filename);
      this.set('content-length', stat.size);
      return this.body = fs.createReadStream(__filename);
    }
    if (this.url === '/exists-encoding') {
      this.set('content-encoding', 'gzip');
      return this.body = new Buffer('gzip');
    }
    if (this.url === '/error') {
      return this.throw(new Error('mock error'));
    }
  });

  before(function (done) {
    app = app.listen(0, done);
  });

  afterEach(mm.restore);

  describe('gzip()', function () {
    it('should work with no options', function () {
      gzip();
    });
  });

  describe('when status 200 and request accept-encoding include gzip', function () {
    it('should return gzip string body', function (done) {
      request(app)
      .get('/string')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('content-encoding', 'gzip')
      .expect('content-length', '46')
      .expect(BODY, done);
    });

    it('should return raw string body if body smaller than minLength', function (done) {
      request(app)
      .get('/small')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('content-length', '14')
      .expect('foo bar string', function (err, res) {
        should.not.exist(err);
        should.not.exist(res.headers['content-encoding']);
        done();
      });
    });

    it('should return raw string body if gzip body bigger than raw body', function (done) {
      mm(options, 'minLength', 10);
      request(app)
      .get('/small')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('content-length', '14')
      .expect('foo bar string', function (err, res) {
        should.not.exist(err);
        should.not.exist(res.headers['content-encoding']);
        done();
      });
    });

    it('should return gzip buffer body', function (done) {
      request(app)
      .get('/buffer')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('content-encoding', 'gzip')
      .expect('content-length', '46')
      .expect(BODY, done);
    });

    it('should return gzip stream body', function (done) {
      request(app)
      .get('/stream')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('Content-Encoding', 'gzip')
      .expect(fs.readFileSync(__filename, 'utf8'),
      function (err, res) {
        should.not.exist(err);
        should.not.exist(res.headers['content-length']);
        done();
      });
    });

    it('should return gzip json body', function (done) {
      request(app)
      .get('/object')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(200)
      .expect('Content-Encoding', 'gzip')
      .expect('content-length', '52')
      .expect({foo: BODY}, done);
    });

    it('should return number body', function (done) {
      request(app)
      .get('/number')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect('content-length', '4')
      .expect(200, function (err, res) {
        should.not.exist(err);
        res.body.should.equal(1984);
        done();
      });
    });
  });

  describe('when status 200 and request accept-encoding exclude gzip', function () {
    it('should return raw body', function (done) {
      request(app)
      .get('/string')
      .set('Accept-Encoding', 'deflate,sdch')
      .expect(200)
      .expect('content-length', '' + BODY.length)
      .expect(BODY,
      function (err, res) {
        should.not.exist(err);
        should.not.exist(res.headers['content-encoding']);
        done();
      });
    });
  });

  describe('when status non 200', function () {
    it('should return 404', function (done) {
      request(app)
      .get('/404')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(404)
      .expect('Not Found', done);
    });

    it('should return 500', function (done) {
      request(app)
      .get('/error')
      .set('Accept-Encoding', 'gzip,deflate,sdch')
      .expect(500)
      .expect('Internal Server Error', done);
    });
  });
});


index.js
/**!
 * koa-gzip - index.js
 *
 * Copyright(c) 2014
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

"use strict";

/**
 * Module dependencies.
 */

var zlib = require('zlib');
var thunkify = require('thunkify-wrap');

var _gzip = thunkify(zlib.gzip);

module.exports = function (options) {
  options = options || {};
  // https://developers.google.com/speed/docs/best-practices/payload#GzipCompression
  options.minLength = Number(options.minLength) || 150;

  return function *gzip(next) {
    yield *next;

    var body = this.body;

    if (200 !== this.status
        || !body
        || this.acceptsEncodings('gzip') !== 'gzip'
        || this.response.header['content-encoding']) {
      return;
    }

    // TODO: Stream body
    if ('function' === typeof body.pipe) {
      this.set('content-encoding', 'gzip');
      this.remove('content-length');
      this.body = body.pipe(zlib.createGzip());
      return;
    }

    if (typeof body === 'string') {
      body = new Buffer(body);
    } else if (!Buffer.isBuffer(body)) {
      body = new Buffer(JSON.stringify(body));
    }

    if (body.length < options.minLength) {
      return;
    }

    var buf = yield _gzip(body);
    if (buf.length > body.length) {
      return;
    }

    this.set('content-encoding', 'gzip');
    this.body = buf;
  };
};
